### PR TITLE
[build][pytorch] patch PT1.4 for python and awscli version

### PR DESCRIPTION
**Pull request by tejaschumbalkar**
*Issue #, if available:*

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON Checklist
* When creating a PR:
- [ ] I've modified `src/config/build_config.py` in my PR branch by setting `ENABLE_EI_MODE = True` or `ENABLE_NEURON_MODE = True`
* When PR is reviewed and ready to be merged:
- [ ] I've reverted the code change on the config file mentioned above
#### Benchmark Checklist
* When creating a PR:
- [ ] I've modified `src/config/test_config.py` in my PR branch by setting `ENABLE_BENCHMARK_DEV_MODE = True`
* When PR is reviewed and ready to be merged:
- [ ] I've reverted the code change on the config file mentioned above

## Reviewer Checklist
* For reviewer, before merging, please cross-check:
- [ ] I've verified the code change on the config file mentioned above has already been reverted

*Description:*
- patch PT1.4 for python and awscli version

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pytorch/training/docker/1.4.0/py3/Dockerfile.cpu [hunk 1]
```diff
@@ -87,7 +89,7 @@ RUN curl -L -o ~/miniconda.sh https://repo.continuum.io/miniconda/Miniconda3-lat
     Pillow \
     h5py==2.9.0 \
     requests==2.22.0 \
- && conda install -c dglteam -y dgl==0.4.1 \
+ && conda install -c dglteam -y dgl==0.4.3post2 \
```
Comment (tejaschumbalkar) [inline pytorch/training/docker/1.4.0/py3/Dockerfile.cpu:92]: bumped up the dgl version as per the recommendation from the DGL team

File: src/config/build_config.py [hunk 2]
```diff
@@ -5,7 +5,7 @@
 # Do remember to revert it back to False before merging any PR (including NEURON dedicated PR)
 ENABLE_NEURON_MODE = False
 # Frameworks for which you want to disable both builds and tests
-DISABLE_FRAMEWORK_TESTS = []
+DISABLE_FRAMEWORK_TESTS = ["tensorflow", "mxnet"]
```
Comment (tejaschumbalkar) [inline src/config/build_config.py:8]: Temp config. To be reverted after review.

File: test/dlc_tests/ec2/pytorch/inference/test_pytorch_inference.py [hunk 3]
```diff
@@ -110,7 +110,7 @@ def ec2_pytorch_inference(image_uri, processor, ec2_connection, region):
 @pytest.mark.integration("telemetry")
 @pytest.mark.model("N/A")
 @pytest.mark.parametrize("ec2_instance_type", PT_EC2_SINGLE_GPU_INSTANCE_TYPE, indirect=True)
-def test_pytorch_inference_telemetry_gpu(pytorch_inference, ec2_connection, gpu_only, ec2_instance_type):
+def test_pytorch_inference_telemetry_gpu(pytorch_inference, ec2_connection, gpu_only, ec2_instance_type, pt15_and_above_only):
```
Comment (tejaschumbalkar) [inline test/dlc_tests/ec2/pytorch/inference/test_pytorch_inference.py:113]: Skipped for PT1.4 as per recommendation from Pytorch team

File: test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py [hunk 4]
```diff
@@ -163,14 +163,13 @@ def test_pytorch_amp(pytorch_training, ec2_connection, gpu_only, ec2_instance_ty
 @pytest.mark.integration("telemetry")
 @pytest.mark.model("N/A")
 @pytest.mark.parametrize("ec2_instance_type", PT_EC2_SINGLE_GPU_INSTANCE_TYPE, indirect=True)
-def test_pytorch_telemetry_gpu(pytorch_training, ec2_connection, gpu_only, ec2_instance_type):
+def test_pytorch_telemetry_gpu(pytorch_training, ec2_connection, gpu_only, ec2_instance_type, pt15_and_above_only):
```
Comment (tejaschumbalkar) [inline test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py:166]: Skipped for PT1.4 as per recommendation from Pytorch team

File: test/dlc_tests/sanity/test_pre_release.py [hunk 5]
```diff
@@ -258,14 +258,12 @@ def _run_dependency_check_test(image, ec2_connection, processor):
 
 @pytest.mark.model("N/A")
 @pytest.mark.parametrize("ec2_instance_type", ["c5.4xlarge"], indirect=True)
```
Comment (tejaschumbalkar) [inline test/dlc_tests/sanity/test_pre_release.py:260]: Temp config. To be added after review.

File: test/dlc_tests/sanity/test_safety_check.py [hunk 6]
```diff
@@ -127,9 +127,6 @@ def _get_latest_package_version(package):
 
 @pytest.mark.model("N/A")
 @pytest.mark.skipif(not is_dlc_cicd_context(), reason="Skipping test because it is not running in dlc cicd infra")
```
Comment (tejaschumbalkar) [inline test/dlc_tests/sanity/test_safety_check.py:129]: Temp config. To be added after review.

File: pytorch/training/docker/1.4.0/py3/Dockerfile.cpu [hunk 7]
```diff
@@ -96,13 +98,18 @@ RUN curl -L -o ~/miniconda.sh https://repo.continuum.io/miniconda/Miniconda3-lat
...
+    "cryptography>3.2" \
+    ruamel-yaml \
     fastai==1.0.59 \
     scipy==1.2.2 \
     smdebug==${SMDEBUG_VERSION} \
-    "sagemaker<2" \
+    "sagemaker>=2,<3" \
     sagemaker-experiments==0.* \
     "sagemaker-pytorch-training<2" \
+    sagemaker-training \
```
Comment (saimidu) [inline pytorch/training/docker/1.4.0/py3/Dockerfile.cpu:112]: `sagemaker-training` is already a part of the requirements of `sagemaker-pytorch-training` and doesn't need to be installed again.

File: pytorch/training/docker/1.4.0/py3/cu101/Dockerfile.gpu [hunk 8]
```diff
@@ -123,9 +125,14 @@ RUN /opt/conda/bin/conda config --set ssl_verify False \
 # custom versions from an S3 bucket. This will need to be removed in the future
 RUN pip install \
     --no-cache-dir smdebug==${SMDEBUG_VERSION} \
-    "sagemaker<2" \
+    "sagemaker>=2,<3" \
+    sagemaker-training \
```
Comment (saimidu) [inline pytorch/training/docker/1.4.0/py3/cu101/Dockerfile.gpu:129]: Please install `sagemaker-pytorch-training`, and not `sagemaker-training`, and remove the sagemaker-pytorch-training installed at the end of the file.

File: test/test_utils/__init__.py [hunk 9]
```diff
@@ -299,7 +285,7 @@ def request_pytorch_inference_densenet(ip_address="127.0.0.1", port="80", connec
         inference_output = json.loads(run_out.stdout.strip("\n"))
         if not (
                 ("neuron" in model_name and isinstance(inference_output, list) and len(inference_output) == 3)
-                or (isinstance(inference_output, dict) and len(inference_output) == 5)
+                or (isinstance(inference_output, list) and len(inference_output) == 5)
```
Comment (saimidu) [inline test/test_utils/__init__.py:288]: Does this work with PT 1.6 and 1.7 as well?

File: pytorch/training/docker/1.4.0/py3/Dockerfile.cpu (hunk 7 above)
Comment (tejaschumbalkar) [reply pytorch/training/docker/1.4.0/py3/Dockerfile.cpu:112]: We still see the error in the test ` ModuleNotFoundError: No module named 'sagemaker_training'`

Checking with the PT team, they recommended to install this explicitly.

File: pytorch/training/docker/1.4.0/py3/cu101/Dockerfile.gpu (hunk 8 above)
Comment (tejaschumbalkar) [reply pytorch/training/docker/1.4.0/py3/cu101/Dockerfile.gpu:129]: See above

File: test/test_utils/__init__.py [hunk 10]
```diff
@@ -96,21 +96,7 @@ def is_below_tf_version(version_upper_bound, image_uri):
...
-    :return: bool True if image_uri has framework version less than version_upper_bound, else False
-    """
-    image_framework_name, image_framework_version = get_framework_and_version_from_tag(image_uri)
-    required_version_specifier_set = SpecifierSet(f"<{version_upper_bound}")
-    return image_framework_name == "mxnet" and image_framework_version in required_version_specifier_set
-
+    return image_framework_name == framework and image_framework_version in required_version_specifier_set
 
 def is_below_pytorch_version(version_upper_bound, image_uri):
     """
```
[the remark this answers is not in the document]
Comment (tejaschumbalkar) [reply test/test_utils/__init__.py]: removed.

**Review by saimidu — COMMENTED** [review]
APPROVED. Please revert temp changes.

**Review by saimidu — APPROVED** [review]

**Pull request merged**